---
title: "SARS-CoV-2 Wastewater Sampling Reports"
nav: "Overview"
author: "BIMSB Bioinformatics Platform"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  variants_csv: ''
  mutations_csv: ''
  coverage_dir: ''
  sample_sheet: ''
  logo: ''
  fun_cvrg_scr: ''
  fun_lm: ''
---

<style>
.pigx_info {
  background-image: url(params$logo); 
}

.notice {
  margin-left: -12px;
  padding-left: 6px;
  border-left: 6px solid #ff9999;
}
</style>

<div id="logo" align="top">
`r knitr::include_graphics(params$logo)`
</div>


```{r libraries and input, echo=F, message=F, warning=F}
library(plotly)
library(htmltools)
library(reshape2)
library(base64url)
library(DT)
library(tidyverse)

df_var <- read.table(params$variants_csv, sep = "\t", header=T)
df_mut <- read.table(params$mutations_csv, sep = "\t", header=T)
```

This pipeline performs mutation analysis of SARS-CoV-2 and reports and quantifies the occurrence of
*variants of concern* (VOC) and signature mutations by which they are characterised.

The visualizations below provide an overview of the evolution of VOCs found
in the analyzed samples across given time points and locations. The abundance values for the variants are derived 
by deconvolution (for details please see the [variant report](https://bimsbstatic.mdc-berlin.de/akalin/PiGx/sars-cov2-ww/example_report/WW_210304_KW1.variantreport_p_sample.html#2_Deconvolution)) 
The frequencies of the mutations are the output of [LoFreq](https://csb5.github.io/lofreq/).

This pipeline is part of
[PiGx](https://bioinformatics.mdc-berlin.de/pigx), a collection of
highly reproducible genomics pipelines
[developed](https://github.com/BIMSBbioinfo/pigx_sarscov2_ww) by the
[Bioinformatics & Omics Data Science
platform](https://bioinformatics.mdc-berlin.de/) at the Berlin
Institute of Medical System Biology (BIMSB).

# Variants Of Concern

These plots provide an overview of the relative frequencies of
identified variants of concern (VOC) of SARS-CoV-2 at specific
wastewater sampling locations over time.

## Frequency of variants of concern over time for each location

These plots show the relative frequency of detected variants of
concern in samples at specific wastewater sampling locations, and how
the frequencies change over time.

Since not all variants of concern are listed, the relative
frequencies at a given location and time will not necessarily add up
to one.
```{r sample_quality_score, echo=FALSE, message=FALSE, warning=FALSE}
source(params$fun_cvrg_scr) # TODO given as input? or only the scripts dir? similar variant_report

coverages.df <- get_genome_cov( params$coverage_dir )
if ( !(length(coverages.df) == 0)){
    good_samples.df <- coverages.df %>% filter( coverage >= 95) # TODO stringecy should be given by user, maybe even by visualization
    bad_samples.df <- coverages.df %>% filter( coverage < 95)
    
    # table displaying bad samples which are not further inlcuded 
    cat("\n Samples without sufficient genome coverage. They aren't included in the visualization or          the linear regression") # TODO refine this text 
    # TODO refine column names --> "coverage in percent", 
    DT::datatable(bad_samples.df,
                  extensions = c('FixedColumns', 'Scroller'),
                  options = list(fixedColumns = TRUE, 
                             scrollY = 180,
                             scrollX = TRUE,
                             scroller = TRUE,
                             dom = 'Bfrtip'),
                  filter = "bottom")
} else {
    cat("\n No coverage values found.")
}
```
```{r filter_plot_frames_samplescore}

approved_var_plot <- df_var %>% filter( samplename  %in% good_samples.df$samplename )
approved_mut_plot <- df_mut %>% filter( samplename  %in% good_samples.df$samplename )
```
```{r color_palette}
Variant_colors <- c("#8a3874", "#4d3da1", "#62c742", "#13317a", "#86e5a3", "#6f69e5", "#4cdab2", "#999ce4", "#2ce6f3",
"#6638c6", "#005f29", "#5d73cc", "#616524", "#a8a041", "#e9f600", "#3eaca8", "#62a273", "#4ea5d0", "#655c95", "#67a04e")
```
```{r variants, echo=F, message=F, warning=F}

melted <- reshape2::melt(df_var,
                         id.vars = c("dates", "samplename", "location_name",
                                     "coordinates_lat", "coordinates_long"))
locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       mutate (value = round (value, 4)) %>%
                       plot_ly (type = 'scatter',
                                mode = 'lines+markers',
                                stackgroup = 'one',
                                showlegend = TRUE,
                                name = ~variable,
                                x = ~dates,
                                y = ~value,
                                color = ~variable,
                                colors = Variant_colors) %>%
                       layout (title = location,
                               showlegend = TRUE,
                               yaxis = list(title = "Frequency"))
               }))
```


## Variants of concern per date and location

This plot visualizes the development of local outbreaks and the
proportions of identified VOCs in that area.  The gray marker at every
sampling location is overlayed by a colored disks for every variant.

<div class="notice">
Locations of wastewater processing plants have been generated arbitrarily and do 
not correspond to actual locations.
</div>

Use the slider to select a specific date or hit the *Play* button to
display all snapshots successively.  Click on a variant in the legend
to toggle its visibility in the map; double-click to view only the
selected variant.

```{r maps for variants, echo=F, message=FALSE, warning=FALSE, out.width="900px", out.height="900px"}

melted <- melt(df_var,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

locations <- unique(melt(df_var,
                         id.vars = c("coordinates_lat",
                                     "coordinates_long")))

plot_ly (type = 'scattermapbox',
         mode = 'markers',
         colors = Variant_colors) %>%
    add_trace(data = locations,
              size = 1,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(data = melted,
              lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%
    layout (autosize = TRUE,
            mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```

## Raw data VOC frequencies per sample
```{r, include = F}
variant_freqs <- readLines(params$variants_csv) %>% paste0(collapse="\n") %>% base64_urlencode()
``` 
[Download Variant_frequencies.csv](`r sprintf('data:text/csv;base64,%s', variant_freqs)`)

# Signature Mutations

Signature mutations are characterizing variants of concern (VOC) of
SARS-CoV-2.  The following plots provide an overview of detected
mutations in different locations, and how their relative frequency
changes over time.

Note that these plots only display signature mutations which where
reported for VOCs.

## Mutations by location

These plots show the relative frequency of detected signature
mutations in samples at specific wastewater sampling locations, and
how the frequencies change over time.

```{r mutations, echo=F, message=F, warning=F}
melted <- melt(df_mut,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE)

# sort by dates
idx <- order(melted$dates)
melted <- melted[idx, ]

locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       plot_ly(type = 'scatter',
                               mode = 'line',
                               name = ~variable,
                               x = ~dates,
                               y = ~value) %>%
                       layout(title = location,
                              yaxis = list(title = "Frequency"))
               }))
```

## Signature mutations by location

This plot visualizes the development of local outbreaks on the level
of single mutations.  The gray marker at every sampling location is
overlayed by a colored disks for each mutation.

<div class="notice">
Locations of wastewater processing plants have been generated arbitrarily and do 
not correspond to actual locations.
</div>

Use the slider to select a specific date or hit the *Play* button to
display all snapshots successively.

```{r mutations for locations and date, echo=F, message=F, warning=F, out.width="900px", out.height="900px"}
melted <- melt(df_mut,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

locations <- unique(melt (df_mut,
                          id.vars = c("coordinates_lat",
                                      "coordinates_long")))

plot_ly (type = 'scattermapbox',
         mode = 'markers') %>%
    add_trace(size = 1,
              data = locations,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(data = melted,
              lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%

    layout (title = ~dates,
            autosize = TRUE,
            mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```

## Raw data signature mutation frequencies per sample
```{r, include = F}
mutation_freqs <- readLines(params$mutations_csv) %>% paste0(collapse="\n") %>% base64_urlencode()
``` 
[Download Mutation_frequencies.csv](`r sprintf('data:text/csv;base64,%s', mutation_freqs)`)

# Detailed per-sample reports

For every sample three reports are generated:

+ a QC report reporting general statistics and amplicon coverage,

+ a variant report including tables summarizing the mutation
  calling and the deconvolution results for the abundance of VOCs,

+ a taxonomic classification report including a pie chart showing the
  analysis of the unaligned reads.

The reports for each sample can be accessed here:

```{r generate overview, warning=F, message=F}
library(data.table)
library(DT)

sample_sheet <- fread(params$sample_sheet)
sample_names <- sample_sheet$name
reports <- list(list("suffix" = ".variantreport_p_sample.html",
                     "name"   = "variant report"),
                list("suffix" = ".qc_report_per_sample.html",
                     "name"   = "QC report"),
                list("suffix" = ".taxonomic_classification.html",
                     "name"   = "taxonomic classification"))

df <- as.data.frame(select(sample_sheet, name, location_name, date))

links <- lapply(sample_names, function (sample) {
    as.vector(lapply(reports, function (report) {
        paste0("<a href=", sample, report$suffix, ">", report$name, "</a>")
    }))
})
df$reports <- links

datatable(df, escape=FALSE)
```
