---
title: "SARS-CoV-2 Wastewater Sampling Reports"
nav: "Overview"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  variants_csv: ''
  mutations_csv: ''
  sample_sheet: ''
  logo: ''
---

<style>
#logo { position: relative; }
#logo img {
  top: 75px;
  left: 50px;
  position: fixed;
  width: 125px;
}
</style>

<div id="logo" align="top">
`r knitr::include_graphics(params$logo)`
</div>


```{r libraries and input, echo=F, message=F, warning=F}
library(plotly)
library(htmltools)
library(reshape2)

df_var <- read.table(params$variants_csv, sep = "\t", header=T)
df_mut <- read.table(params$mutations_csv, sep = "\t", header=T)
```

This pipeline performs mutation analysis of SARS-CoV-2 and reports and quantifies the occurrence of
*variants of concern* (VOC) and signature mutations by which they are characterised.

# Overview

For each sample three reports are generated:

+ a QC report reporting general statistics and amplicon coverage,

+ a variant report including tables summarizing the mutation
  calling and the deconvolution results for the abundance of VOCs,

+ a taxonomic classification report including a pie chart showing the
  analysis of the unaligned reads.

The reports for each sample can be accessed here:

```{r generate overview, warning=F, message=F}
library(data.table)
library(dplyr)
library(DT)
library(reshape2)

sample_sheet <- fread(params$sample_sheet)
sample_names <- sample_sheet$name
reports <- list(list("suffix" = ".variantreport_p_sample.html",
                     "name"   = "variant report"),
                list("suffix" = ".qc_report_per_sample.html",
                     "name"   = "QC report"),
                list("suffix" = ".taxonomic_classification.html",
                     "name"   = "taxonomic classification"))

df <- as.data.frame(select(sample_sheet, name, location_name, date))

links <- lapply(sample_names, function (sample) {
    as.vector(lapply(reports, function (report) {
        paste0("<a href=", sample, report$suffix, ">", report$name, "</a>")
    }))
})
df$reports <- links

datatable(df, escape=FALSE)
```


The visualizations below provide an overview of the evolution of VOCs found
in the analyzed samples across given time points and locations. The abundance values for the variants are derived 
by deconvolution (for details please see the [variant report](https://bimsbstatic.mdc-berlin.de/akalin/PiGx/sars-cov2-ww/example_report/WW_210304_KW1.variantreport_p_sample.html#22_Deconvolution)) 
The frequencies of the mutations are the output of [LoFreq](https://csb5.github.io/lofreq/).


# Virus Strains

These plots provide an overview of the relative frequencies of
identified variants of concern (VOC) of SARS-CoV-2 at specific
wastewater sampling locations over time.

## Frequency of virus strains over time for each location

These plots show the relative frequency of detected variants of
concern in samples at specific wastewater sampling locations, and how
the frequencies change over time.

Since not all strains are listed as variants of concern, the relative
frequencies at a given location and time will not necessarily add up
to one.

```{r variants, echo=F, message=F, warning=F}

melted <- reshape2::melt(df_var,
                         id.vars = c("dates", "samplename", "location_name",
                                     "coordinates_lat", "coordinates_long"))
melted$dates <- as.Date(melted$dates)
locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       mutate (value = round (value, 4)) %>%
                       plot_ly (type = 'scatter',
                                mode = 'line',
                                stackgroup = 'one',
                                showlegend = TRUE,
                                x = ~dates,
                                y = ~value,
                                color = ~variable) %>%
                       layout (title = location)
               }))
```


## Virus strains per date and location

This plot visualizes the development of local outbreaks and the
proportions of identified VOCs in that area.  The gray marker at every
sampling location is overlayed by a colored disks for every variant.
The diameter of these disks corresponds to the relative frequency of
the given variant at the specific site.

Use the slider to select a specific date or hit the *Play* button to
display all snapshots successively.  Click on a variant in the legend
to toggle its visibility in the map; double-click to view only the
selected variant.

```{r maps for variants, echo=F, message=FALSE, warning=FALSE}

melted <- melt(df_var,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

plot_ly (data = melted,
         type = 'scattermapbox',
         mode = 'markers') %>%
    add_trace(size = 1,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              name = ~variable,
              legendgroup = ~variable,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%
    layout (width = 1000,
            mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```


# Signature Mutations

Signature mutations are characterizing variants of concern (VOC) of
SARS-CoV-2.  The following plots provide an overview of detected
mutations in different locations, and how their relative frequency
changes over time.

Note that these plots only display signature mutations which where
reported for VOCs.

## Mutations by location

These plots show the relative frequency of detected signature
mutations in samples at specific wastewater sampling locations, and
how the frequencies change over time.

```{r mutations, echo=F, message=F, warning=F}
melted <- reshape2::melt(df_mut,
                         id.vars = c("dates", "samplename", "location_name",
                                     "coordinates_lat", "coordinates_long"))
melted$dates <- as.Date(melted$dates)

# sort by dates
idx <- order(melted$dates)
melted <- melted[idx, ]

locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       plot_ly(type = 'scatter',
                               mode = 'line',
                               name = ~variable,
                               x = ~dates,
                               y = ~value) %>%
                       layout(title = location)
               }))
```

## Signature mutations by location

This plot visualizes the development of local outbreaks on the level
of single mutations.  The gray marker at every sampling location is
overlayed by a colored disks for each mutation.  The diameter of these
disks corresponds to the relative frequency of the given mutation at
the specific site.

Use the slider to select a specific date or hit the *Play* button to
display all snapshots successively.

```{r mutations for locations and date, echo=F, message=F, warning=F}
melted <- melt(df_mut,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

plot_ly (data = melted,
         type = 'scattermapbox',
         mode = 'markers') %>%
    add_trace(size = 1,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              name = ~variable,
              legendgroup = ~variable,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%

    layout (title = ~dates,
            width = 1000,
            mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```
