---
title: "Timecourse"
nav: "Timecourse"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  variants_csv: ''
  mutations_csv: ''
---

```{r libraries and input, echo=F, message=F, warning=F}
library(plotly)
library(htmltools)
library(reshape2)

df_var <- read.table(params$variants_csv, sep = "\t", header=T)
df_mut <- read.table(params$mutations_csv, sep = "\t", header=T)
```
This pipeline performs mutation analysis of SARS-CoV-2 and reports and quantifies the occurrence of
variants of concern (VOC) and signature mutations by which they are characterised. For each sample three reports are 
generated:  
* a QC-report reporting general statistics and amplicon coverage
* a variant-report which including tables summarizing the mutation calling and the deconvolution results for the abundance of VOCs
* a taxonomic classification report including a pie chart showing the analysis of the unaligned reads

The visualizations below provide an overview of the evolution of VOCs found
in the analyzed samples across given time points and locations. The abundance values for the variants are derived 
by deconvolution (for details please see the [variant report](https://bimsbstatic.mdc-berlin.de/akalin/PiGx/sars-cov2-ww/example_report/WW_210304_KW1.variantreport_p_sample.html#22_Deconvolution)) 
The frequencies of the mutations are the output of [LoFreq](https://csb5.github.io/lofreq/).


# VARIANTS

## Frequency of virus strains over time for each location
```{r variants, echo=F, message=F, warning=F}

melted <- reshape2::melt(df_var,
                         id.vars = c("dates", "samplename", "location_name",
                                     "coordinates_lat", "coordinates_long"))
melted$dates <- as.Date(melted$dates)
locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       mutate (value = round (value, 4)) %>%
                       plot_ly (type = 'scatter',
                                mode = 'line',
                                stackgroup = 'one',
                                showlegend = TRUE,
                                x = ~dates,
                                y = ~value,
                                color = ~variable) %>%
                       layout (title = location)
               }))
```


## Virus strains per date and location

On this map you can see the relative frequencies of identified variants of concern (VOC) of SARS-CoV-2 at specific 
wastewater sampling locations over time. This plot 
visualizes the development of local outbreaks and the proportions of identified VOCs in that area.  

Use the slider to select a specific date or hit the Play button to display all snapshots successively. By 
double-clicking on the variants displayed on the legend the chosen variants can be disabled.

```{r maps for variants, echo=F, message=FALSE, warning=FALSE}

melted <- melt(df_var,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

plot_ly (data = melted,
         type = 'scattermapbox',
         mode = 'markers') %>%
    add_trace(size = 1,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              name = ~variable,
              legendgroup = ~variable,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%
    layout (mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```


# MUTATIONS

## mutations  over time for each location 
```{r mutations, echo=F, message=F, warning=F}
melted <- reshape2::melt(df_mut,
                         id.vars = c("dates", "samplename", "location_name",
                                     "coordinates_lat", "coordinates_long"))
melted$dates <- as.Date(melted$dates)

# sort by dates
idx <- order(melted$dates)
melted <- melted[idx, ]

locations <- unique(melted$location_name)

tagList(lapply(locations,
               function (location) {
                   subset(melted, location_name == location) %>%
                       group_by(dates) %>%
                       arrange(variable) %>%
                       plot_ly(type = 'scatter',
                               mode = 'line',
                               name = ~variable,
                               x = ~dates,
                               y = ~value) %>%
                       layout(title = location)
               }))
```

## mutations on the map for each data/location

This map shows the frequency values of signature mutations which are characterising variants of concern (VOC) of SARS-CoV-2 for
samples at specific wastewater sampling locations over time. The diameter of the circles displays the mutation frequency. 
Colors of overlapping variants are adding up. This plot visualizes the development of local outbreaks on the level of 
single mutations. So far the map can only display signature mutations which where reported for VOCs.

Use the slider to select a specific date or hit the Play button to display all snapshots successively.

```{r mutations for locations and date, echo=F, message=F, warning=F}
melted <- melt(df_mut,
               id.vars = c("dates", "samplename", "location_name",
                           "coordinates_lat", "coordinates_long"),
               rm.NA = TRUE) %>%
    mutate (value = round (value, 4))

plot_ly (data = melted,
         type = 'scattermapbox',
         mode = 'markers') %>%
    add_trace(size = 1,
              lon = ~coordinates_long,
              lat = ~coordinates_lat,
              name = ~variable,
              legendgroup = ~variable,
              opacity = 0.1,
              showlegend = FALSE,
              hoverinfo = 'none',
              marker = list(sizemode = 'diameter',
                            color = 'rgb(150, 150, 150)',
                            opacity = 0.1)) %>%
    add_trace(lat = ~coordinates_lat,
              lon = ~coordinates_long,
              legendgroup = ~variable,
              showlegend = TRUE,
              frame = ~dates,
              size = ~value,
              color = ~variable,
              hoverinfo = 'text+name',
              hovertext = ~paste0("Frequency: ", value),
              marker = list (sizemode = 'diameter')) %>%

    layout (title = ~dates,
            mapbox = list (zoom = 5.5,
                           center = list(lat = ~median(coordinates_lat, na.rm=TRUE),
                                         lon = ~median(coordinates_long, na.rm=TRUE)),
                           style = 'open-street-map'))
```
