---
title: "taxonomic classification"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
#output: html_document
params:
  @PARAMS_SAMPLE_NAME@

---

Read config.yml for the locations of sample sheet, pipeline output and krakenDir. 
```{r, child="_settings.Rmd"}
```

```{r input, echo=FALSE}
#input 
sampleName <- params$sample_name
krakenDir <- config$kraken_dir

krakenReportFile <- file.path(krakenDir, paste0(sampleName, '_classified_unaligned_reads.txt'))
kronaFile <- file.path(site_dir, paste0(sampleName, '_krona_report.html'))
```


# Kraken2 report on unaligned reads from the Sars-Cov2 enriched samples

```{r print sample name,echo=F}
print(paste( "Sample: ",sampleName))
```


The SARS-CoV-2 enriched wastewater samples are aligned to the virus genome. The unaligned reads, that are left are aligned with Kraken2 against the [database PlusPFP](https://benlangmead.github.io/aws-indexes/k2). This report is an overview over the species found.  

A table with all the species found: 

* Column explanation:   
  + Percentage of fragments covered by the clade rooted at this taxon  
  + Number of fragments covered by the clade rooted at this taxon  
  + Number of fragments assigned directly to this taxon  
  + A rank code, indicating (U)nclassified, (R)oot, (D)omain, (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies.     Taxa that are not at any of these 10 ranks have a rank code that is formed by using the rank code of the closest ancestor rank         with a number indicating the distance from that rank. E.g., "G2" is a rank code indicating a taxon is between genus and species         and the grandparent taxon is at the genus rank.  
  + NCBI taxonomic ID number  
  + Indented scientific name  

```{r plot table, echo=FALSE}
table<- read.table(krakenReportFile,sep = "\t")
header_table <- c("percent","num_fragments","num_fragments_taxon","rank","NCBI_ID","scientific_name")
colnames(table)<-header_table
datatable(table)
```

# visualization of all unaligned reads are available with Krona:

The data can be explored with a multi-layered pie chart, which was created with [KronaTools](https://github.com/marbl/Krona/wiki). It is an interactive chart of all the classificated species found with Kraken2.  


[Krona graph of this sample](`r kronaFile`)