---
title: "Taxonomic Classification of Unaligned Reads"
nav: "Taxonomy"
author: "BIMSB Bioinformatics Platform"
date: '`r format(as.POSIXct(if ("" != Sys.getenv("SOURCE_DATE_EPOCH")) { as.numeric(Sys.getenv("SOURCE_DATE_EPOCH")) } else { Sys.time() }, origin="1970-01-01"), "%Y-%m-%d %H:%M:%S")`'
params:
  sample_name: ''
  site_dir: ''
  kraken_file: ''
  krona_file: ''
  logo: ''
---
<style>
.dropdown-menu {
  max-height: 200px;
  overflow-y: scroll;}
</style>

<div id="logo" align="top">
`r knitr::include_graphics(params$logo)`
</div>

<sample>`r params$sample_name`</sample>

```{r libs, echo=FALSE}
suppressMessages(suppressWarnings(library(DT)))
library(knitr)
```

This metagenomic analysis report provides an overview of the taxa that were classified from the set of unaligned reads 
of the given sample. The analysis of unaligned reads can provide insight about the alignment quality, cleanliness 
of the sample, possible species with high sequence similarity which could introduce alignment bias and more.  
<font size="3"> *Note: Because of algorithmic differences between the classification and the sequence alignment, a high 
number of fragments assigned to the species of interest (in this case SARS-CoV-2) could be a result of differences in 
thresholds for sequence similarity. In this case, further adjustment of the alignment tool and/or the classifier are
recommended.* </font>  

**Note about value comparison:**  
The number of aligned and unaligned reads of this sample can be found [in the QC overview table](../overview_QC.csv).
Kraken2 reports the classification of fragments. Therefore paired-end read data, the number of fragments is half 
the number of unaligned reads. With single-end read data, the number of fragments should equal the number 
of unaligned reads. 


# Krona Visualization

This interactive multi-layered pie chart displays the proportions of the classified taxons found in the sample. Shown are the taxonomy classes as nested sectors, with decreasing hierarchy from the center towards the periphery. The input values can be found in the table below. 

<iframe id="krona" style="min-height: 100vh" width="100%" onload="this.height = this.contentWindow.document.body.scrollHeight;" src="./`r basename(params$krona_file)`"></iframe>

<font size="3"> 
**Magnitude:** number of fragments covered by currently chosen supergroup displayed in the white middle of the circle   
**Unassigned magnitude:** number of fragments that have no other classification than “root”   
**Count:** number of taxa covered by the current supergroup   
**Unassigned:** number of taxa that have no other ancestor than “root” </font>


# Raw classification report

The table below lists the classified taxa (**Taxon. class**), its NCBI taxonomic ID (**NCBI tax. ID**), taxonomic rank 
(**Rank:** **U**nclassified, **R**oot, **D**omain, **K**ingdom, **P**hylum, **C**lass, **O**rder, **F**amily, **G**enus, 
or **S**pecies ) and the number of fragments that was assigned uniquely to this taxa (Number of assigned fragments). In 
case of higher taxa covering multiple lower ones, the **"Number of covered fragments”** is the sum of all fragments that 
are covered by it (e.g. “Viruses” covers all virus species). In addition, it is shown which proportion of all fragments 
are classified by each taxon (**Propotion in %**) 
  
```{r plot table, echo=FALSE}
table<- read.table(params$kraken_file, sep = "\t")
colnames(table) <- c("Proportion in %", "Number of covered fragments", "Number of assigned fragments", "Rank", "NCBI tax. ID", "Taxon. class")
datatable(table, filter="bottom")
```

Example on interpretation of the counting:   
```{r plot example_table, echo = FALSE}
df <- data.frame( "Proportion in %" = c(40,30),
                  "Number of covered fragments" = c(1000,800),
                  "Number of assigned fragments" = c(200,800),
                  Rank = c("Domain","Species"), 
                  "Taxonomic class" =c("Virus","Sars-CoV-2"), check.names = FALSE
)
kable(df, align = 'l')
```

Read:

* 40% of all analysed fragments are related to Virus and more precisely  
* 30% of all analysed fragments are related to one species - Sars-CoV-2  
* 1000 fragments are related to Virus, but only 200 fragments have no further classification but “Virus”  
* 800 of the virus-related fragments could be classified further down to species level and could all classified as SARS-CoV-2 
