title: Supplementary Materials
---

These are supplementary files for the PiGx paper.


## PiGx scRNA-seq

The HTML report output of the scRNA-seq pipeline from the analysis of
the single cell sequencing dataset from (Hu et al. 2017) [can be
accessed
here](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/scrnaseq/mm10.scRNA-Seq.report.html).


## PiGx ChIP-seq

The HTML report output of the ChIP-seq pipeline from the analysis of
the ChIP-seq datasets from Hon. et al [can be accessed here](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/chipseq/ChIP_Seq_Report.html).


## PiGx RNA-seq

Below are links for the HTML reports for each differential expression 
analysis performed using STAR-based gene counts for the use-case 
RNA-seq datasets:

- [tet2_diff_day3](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_diff_day3.html)

- [tet2_diff_day6](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_diff_day6.html)

- [tet2_vs_WT_day0](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day0.html)

- [tet2_vs_WT_day3](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day3.html)

- [tet2_vs_WT_day6](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day6.html)

- [WT_diff_day3](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/WT_diff_day3.html)

- [WT_diff_day6](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/WT_diff_day6.html)


## PiGx BS-seq

The supplementary results for the BS-seq pipeline from the analysis of
the RNA-seq datasets from Hon. et al consist of three reports in HTML
format; We have highlighted the most important figures from the study.

Each HTML document represents a stand-alone summary of some the key
findings for the given sample, along with a description of the key
operations carried out by the pipeline.

- [Tet2 deletion (biological replicate 1) report](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/tet2_se_bt2.sorted.deduped_mm10_final.html)

- [Wild-type (biological replicate 1) report](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/WT_se_bt2.sorted.deduped_mm10_final.html)

- [Differential methylation, biological replicate 1, (tet2-/-) - WT](http://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/diffmeth-report.0vs1.html)

# About PiGx

PiGx is [free
software](https://www.fsf.org/about/what-is-free-software) under the
[GNU General Public License](https://www.gnu.org/licenses/gpl.html)
(version 3 or, at your option, any later version).  You can [get the
complete source code here](https://github.com/BIMSBbioinfo/pigx).
Your contributions are welcome!

Consider [subscribing here to infrequent announcements regarding
PiGx](https://groups.google.com/forum/#!forum/pigx-announcements/join),
such as release announcements.
