<!DOCTYPE html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" /><meta http-equiv="Content-Language" content="en" /><meta name="author" content="BIMSB Bioinformatics" /><meta name="viewport" content="width=device-width" /><title>Supplementary Materials</title><link rel="stylesheet" media="screen" type="text/css" href="css/screen.css" /><link rel="shortcut icon" href="https://bioinformatics.mdc-berlin.de/pigx/favicon.ico" /></head><body id="top"><div id="index" align="center"><img alt="logo" src="images/logo.svg" width="30%" height="30%" /></div><div id="page"><h1>Supplementary Materials</h1><p>These are supplementary files for the PiGx paper.  All reports were
generated with PiGx 0.0.3 as installed with <a href="https://gnu.org/software/guix">GNU
Guix</a> at version
v0.14.0-7054-g5149aeb7e (git commit
5149aeb7e62cf62398b55be38469cd28c25d8d7d).</p><p>We provide a <a href="https://hub.docker.com/r/bimsbbioinfo/pigx/">generated Docker image on
Dockerhub</a>.  The tag
<code>bimsbbioinfo/pigx:publication</code> corresponds to the image at the above
Guix version, which was used for the publication.</p><p>We also provide a <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/pigx-singularity-image.squashfs">SquashFS file system
image</a>,
which can be used with systems like Singularity.</p><p>The SquashFS image was created with this command:</p><pre><code>guix pack pigx            \
          -f squashfs     \
          -S /bin=bin     \
          -S /lib=lib     \
          -S /share=share \
          glibc-utf8-locales tzdata coreutils bash</code></pre><p>This command was used for generating the Docker image:</p><pre><code>guix pack pigx            \
          -f docker       \
          -C none         \
          -S /bin=bin     \
          -S /lib=lib     \
          -S /share=share \
          glibc-utf8-locales tzdata coreutils bash</code></pre><h2>PiGx scRNA-seq</h2><p>The HTML report output of the scRNA-seq pipeline from the analysis of
the single cell sequencing dataset from (Hu et al. 2017) <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/scrnaseq/mm10.scRNA-Seq.report.html">can be
accessed
here</a>.</p><p>An archive containing a sample sheet, settings file, and instructions
on how to reproduce the report <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/scrnaseq/scrnaseq-notes.tar.gz">can be downloaded
here</a>.</p><h2>PiGx ChIP-seq</h2><p>The HTML report output of the ChIP-seq pipeline from the analysis of
the ChIP-seq datasets from Hon. et al <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/chipseq/ChIP_Seq_Report.html">can be accessed here</a>.</p><p>An archive containing a sample sheet, settings file, and instructions
on how to reproduce the report <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/chipseq/chipseq-notes.tar.gz">can be downloaded
here</a>.</p><h2>PiGx RNA-seq</h2><p>Below are links for the HTML reports for each differential expression
analysis performed using STAR-based gene counts for the use-case
RNA-seq datasets:</p><ul><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_diff_day3.html">tet2_diff_day3</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_diff_day6.html">tet2_diff_day6</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day0.html">tet2_vs_WT_day0</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day3.html">tet2_vs_WT_day3</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/tet2_vs_WT_day6.html">tet2_vs_WT_day6</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/WT_diff_day3.html">WT_diff_day3</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/WT_diff_day6.html">WT_diff_day6</a></p></li></ul><p>An archive containing a sample sheet, settings file, and instructions
on how to reproduce the report <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/rnaseq/rnaseq-notes.tar.gz">can be downloaded
here</a>.</p><h2>PiGx BS-seq</h2><p>The supplementary results for the BS-seq pipeline from the analysis of
the RNA-seq datasets from Hon. et al consist of three reports in HTML
format; We have highlighted the most important figures from the study.</p><p>Each HTML document represents a stand-alone summary of some the key
findings for the given sample, along with a description of the key
operations carried out by the pipeline.</p><ul><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/tet2_se_bt2.sorted.deduped_mm10_final.html">Tet2 deletion (biological replicate 1) report</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/WT_se_bt2.sorted.deduped_mm10_final.html">Wild-type (biological replicate 1) report</a></p></li><li><p><a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/diffmeth-report.0vs1.html">Differential methylation, biological replicate 1, (tet2-/-) - WT</a></p></li></ul><p>An archive containing a sample sheet, settings file, and instructions
on how to reproduce the report <a href="https://bimsbstatic.mdc-berlin.de/akalin/PiGx/supplementary_material/bsseq/bsseq-notes.tar.gz">can be downloaded
here</a>.</p><h1>About PiGx</h1><p>PiGx is <a href="https://www.fsf.org/about/what-is-free-software">free
software</a> under the
<a href="https://www.gnu.org/licenses/gpl.html">GNU General Public License</a>
(version 3 or, at your option, any later version).  You can <a href="https://github.com/BIMSBbioinfo/pigx">get the
complete source code here</a>.
Your contributions are welcome!</p><p>Consider <a href="https://groups.google.com/forum/#!forum/pigx-announcements/join">subscribing here to infrequent announcements regarding
PiGx</a>,
such as release announcements.</p><footer><img alt="decoration" class="end-decoration" src="images/end.svg" /><p>Made by the Akalin lab with love at the Berlin Institute of Medical Systems Biology in 2018</p></footer></div></body>