---
title: "ChIP-seq Sample Report"
output:
  html_document:
    highlight: kate
    theme: spacelab
    toc: yes
    toc_float: yes
always_allow_html: yes

output:
  html_document: default
  
params:
    annotation: '../Data/Annotation.rds'
    path_peaks: '../Data/Peaks1_qsort.narrowPeak'
    path_bw:    '../Data/ChIP1.bw'
    
---

```{r setup, include=FALSE, cache=FALSE, fig.height=10, fig.width=10}
knitr::opts_chunk$set(echo     = FALSE, 
                      cache    = FALSE, 
                      messages = FALSE, 
                      warning  = FALSE,
                      autodep  = TRUE, 
                      error    = FALSE)

```


```{r libraries, include=FALSE}
library(GenomicRanges)
library(data.table)
library(rtracklayer)
library(genomation)
```

```{r input, include=FALSE}
source('../Scripts/Prepare_Annotation.R')
annotation = list(
        gtf = '../Data/mm9.chr19.gtf.gz',
        cpg_island = '../Data/mm9.chr19.cpg.bed')
prepare_Annotation(annotation, outpath='../Data/Annotation.rds', scriptdir = '../Scripts/')
annot = readRDS('../Data/Annotation.rds')
```


# Genomic Distribution


# Signal Distribution Around Genomic Features


# Peak Comparison


# Peak Combination Statistics

# 

