---
title: "PiGx-scRNAseq Report"
author: "BIMSB Bioinformatics Platform"
date: "`r date()`"
output:
  html_document:
    df_print: paged
params:
  covariates: time, replicate
  outFile: test.scRNA.html
  sceRdsFile: /data/local/buyar/pigx/scrna/integrate_report/results_test3/Mapped/hg19.SingleCellExperiment.RDS
  workdir: .
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
knitr::opts_knit$set(root.dir = params$workdir)
library(cowplot)
library(DT)
library(SingleCellExperiment)
library(data.table)
```

# Description

PiGx scRNAseq 

This report was generated with PiGx-scRNAseq version 0.0.1.

# Input Settings
```{r printInputSettings}

sceRdsFile <- params$sceRdsFile
workdir <- params$workdir
outFile <- params$outFile
covariates <- gsub(' ', '', unlist(strsplit(x = params$covariates, split = ',')))

inputParameterDesc <- c('RDS format file containing a SingleCellExperiment object',
                     'Working directory',
                     'Path to HTML report output',
                     'Covariates to use when plotting (PCA and t-SNE)'
                     )
inputParameterValues <- c(sceRdsFile,
                          workdir, 
                          outFile, 
                          paste0(covariates, collapse = ', '))
inputSettings <- data.frame(parameters = inputParameterDesc,
                            values = inputParameterValues,
                            stringsAsFactors = FALSE)
DT::datatable(data = inputSettings,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         pageLength = length(inputParameterValues),
                         dom = 't'))
```


## Content Summary of the SingleCellExperiment Object


First a pre-generated SingleCellExperiment object is read.

This object must minimally have the following `assays`: 

  - cnts (row read counts matrix per gene X cell) 
  - cpm (`cnts` object normalized into counts per million in log2 scale) 
  - scale (centered and scaled version of `cpm`)

And the following `reduced dimension` datasets:

  - PCA: Principle Component Analysis values for each cell (at least top 10 components)
  - TSNE: t-SNE analysis values of for each cell (at least top 2 components)

```{r read-sce-object}
sce <- readRDS(file = sceRdsFile)
print(sce)
```

# Results

```{r cell-stats}
plotCellStats <- function(df, x, y, label_x, label_y, title) {
  p <- ggplot(df, aes_string(x=x, y=y)) +
  geom_boxplot() +
  labs(x = label_x, y = label_y, title = title) + 
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(angle = 90, hjust = 1))
  return(p)
}

plotDesc <- list('nGene' = c('Sample', 
                          'Number of detected genes', 
                          'Number of detected genes per cell'), 
              'max_gene' = c('Sample',
                             'Maximum gene expression per cell',
                             'Maximum gene expression'),
              'mean_expr' = c('Sample',
                              'Average gene expression',
                              'Average gene expression per cell\nfor genes with >0 UMI'))

cellStatsPlots <- lapply(names(plotDesc), function(y){
  p <- plotCellStats(df = as.data.frame(colData(sce)), 
                x = 'sample_id',
                y = y, 
                label_x = plotDesc[[y]][1], 
                label_y = plotDesc[[y]][2],
                title = plotDesc[[y]][3])
  return(p)
})
names(cellStatsPlots) <- paste(lapply(plotDesc, function(x) x[2]))
```

## Cell Statistics {.tabset}

```{r cellStatsPlots, results='asis', echo = FALSE}
for (i in 1:length(cellStatsPlots)) {
  cat("### ",names(cellStatsPlots)[i],"\n")
  print(cellStatsPlots[[i]])
  cat('\n\n')
}
```

## Cell Composition

```{r definePlotPCAFunction, include=TRUE, fig.width=35, fig.height=5}
reducedDimPlot <- function(df, dim1, dim2, title = NULL, color_by = NULL, gradient = FALSE) {
  p <- ggplot2::ggplot(df, aes_string(x = dim1, y = dim2)) +
    geom_point(aes_string(color = color_by), size = 0.5, alpha = 0.5) +
    labs(title = title)
  
  if (gradient == TRUE) {
    p <- p + scale_color_gradient2(midpoint = 3.3)
  }
  return(p)
}
```

```{r CellComposition-PCA}
colData(sce)$log10nGene <- log10(colData(sce)$nGene)
gradientList <- c('log10nGene')
covariateList <- c(covariates, gradientList)
if('CellCyclePhase' %in% colnames(colData(sce))) {
  covariateList <- c(covariateList, 'CellCyclePhase')  
}

pcaData <- as.data.table(cbind(colData(sce), sce@reducedDims$PCA))

pcaPlotList <- lapply(covariateList, function(cov) {
  pL<- lapply(paste0('PC', c(2:8)), function(pc) {
  myTitle <- paste('PC1 vs', pc)
  reducedDimPlot(df = pcaData, 
                 dim1 = 'PC1', 
                 dim2 = pc, 
                 title = myTitle, 
                 color_by = cov, 
                 gradient = cov %in% gradientList)
  })
  names(pL) <- paste0('PC1 vs PC',c(2:8))
  return(pL)
  }
)
names(pcaPlotList) <- covariateList
```


```{r pcaPlotList, results='asis', echo = FALSE}
for(cov in names(pcaPlotList)) {
  cat("### PCA - Colored by",cov," {.tabset} \n")
  for (i in 1:length(pcaPlotList[[cov]])) {
    cat("#### ",names(pcaPlotList[[cov]])[i],"\n")
    print(pcaPlotList[[cov]][[i]])
    cat('\n\n')
  }
}
```


```{r CellComposition-TSNE}
tsneData <- as.data.table(cbind(colData(sce), sce@reducedDims$TSNE))

tsnePlotList <- lapply(covariateList, function(cov) {
  reducedDimPlot(df = tsneData, dim1 = 'V1', dim2 = 'V2', title = 'tSNE_1 vs tSNE_2', color_by = cov)
})
names(tsnePlotList) <- covariateList
```

### t-SNE - colored by different covariates {.tabset}

```{r tsnePlotList, results='asis', echo = FALSE}
for (i in 1:length(tsnePlotList)) {
  cat("#### t-SNE - Colored by ",names(tsnePlotList)[[i]],"\n")
  print(tsnePlotList[[i]])
  cat('\n\n')
}
```

# Session Information
```{r sessionInfo}
sessionInfo()
```
