---
title: "PiGx: BS-seq"
date: '`r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`'
bibliography: reports.bib
params:
  inBam:    '' 
  assembly: ''
  mincov:   10
  minqual:  20
  rdata:    ''
  sessioninfo: TRUE
  references: TRUE
---

```{r include=FALSE}
makeSessioninfo=params$sessioninfo
makeReferences = params$references
```


```{r eval_params, echo=FALSE}

input     <- params$inBam
# output    <- params$outBed
assembly  <- params$assembly
mincov    <- params$mincov
minqual   <- params$minqual
rdatafile <- params$rdata
```


## Methylation Calling Report


<!-- this is the local dir: `r  getwd()` -->
<!-- Here I need to write a short section why we do this -->

```{r, include=FALSE,warning=FALSE}

  suppressPackageStartupMessages(expr = {
  ## load methylKit
    library("methylKit",warn.conflicts = FALSE,quietly = TRUE)
  })
```

### Libraries
```{r, eval=FALSE, include=TRUE}

  ## load methylKit
    library("methylKit")

```

### Parameters
```{r print_params, echo=TRUE, results="asis"}

paramsDesc <- c("Input file",
            "Genome Assembly",
            "Minimum Coverage",
            "Minimim Mapping Quality")

paramsValues <- c( params$inBam,
             params$assembly,
             params$mincov,
             params$minqual)

inputParams <- data.frame(Parameters = paramsDesc,
                          Values = paramsValues,
                          stringsAsFactors = FALSE)
#params
#datatable(inputParams,options = list(dom = 't'),rownames = FALSE)
knitr::kable(inputParams)
```




### Extract Methylation Calls

We first extract the methylation calls from the sequence alignment produced by the bisulfite mapper [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) [@krueger_bismark:_2011] using the `processBismarkAln()` function of [methylKit](https://bioconductor.org/packages/release/bioc/html/methylKit.html) [@methylKit2012]  a package for the the analysis of DNA methylation profiles. In order to ensure data quality we require at least `r mincov` bases coverage and a mapping quality above `r minqual`. 

```{r, message=FALSE}

    ## extract the sample id from sample file 
    # sample_id = unlist(strsplit(basename(input),split = "_"))[1]
    sample_id <- gsub(".bam","",basename(input))

    save_folder <- dirname(rdatafile)

    ## read bam file into methylKit object
    methRaw = processBismarkAln(location = input,
                                sample.id = sample_id,
                                assembly = assembly,
                                mincov = mincov,
                                minqual = minqual,
                                save.context = "CpG",
                                save.folder = save_folder)


  
```
The methylation calls are written to the text file "`r paste0(sample_id,"_CpG.txt")`", 
but to enable direct loading of the `methylRaw` object it is saved for further analysis to "`r basename(rdatafile)`" in the methylation calls directory.
```{r}
    ## Saving object
    save(methRaw,file=normalizePath(rdatafile)) 
```

Here we show some simple statistics as distribution of methylation and coverage.

```{r, fig.show="hold"}


    ## show some statistics of the data  
    par(mfrow = c(1,2))
    getMethylationStats(methRaw,plot = TRUE)
    getCoverageStats(methRaw,plot = TRUE)
    par(mfrow = c(1,1))
    
```

`r if(makeSessioninfo) {"### Session Info"}`

```{r ,include=makeSessioninfo  }
sessionInfo()
```


<!-- this section is used to generate a merged sessioninfo from multiple reports -->
```{r, include=FALSE }
session_file <- tempfile(pattern = 'session',tmpdir = ".",fileext = '.rds')
sess <- sessionInfo()
saveRDS(sess,file = session_file)
```


`r if(makeReferences) {"### References"}`

********