####################################################################
#
# Rules to be applied after mapping reads with Bismark
# ====================================================

# These rules generate a report, such that the user can 
# see wich commands generated the output and 
# can recreate data with modified parameters if not happy with the results.
# 
# 

#
#
# Dependencies:
# - python-rp2, pandoc, methylKit(r), DT(r), genomation(r),
#   annotationhub(r), rtracklayer(r)
# 


    # TODO: 
    # - we need to define a fixed location of TEMPLATEDIR and 
    #   SCRIPTDIR in the config file
####################################################################


"""
This is how the minimal config file should look like, in order to inlcude this snakefile


{
    "reference": {
        "ce10": ["/home/agosdsc/Base/Genomes/ce10/"]
    },
    "samples": {
        "A":["A"],
        "B":["B"]
    },
    "units": {
        "A":
            ["path/to/A_R1.fastq.gz", "path/to/A_R2.fastq.gz"],
        "B":
            ["path/to/B.fastq.gz"]
    },
    "treatment": {
        "A": 1,
        "B": 0
    },
    "params_bam_methCall": {
        "mincov": 10,
        "minqual": 20
    }     
}
"""

## show base dir
# print(workflow.basedir)

import os
## show working dir
# print(os.getcwd())



## show current dir
# print(srcdir(""))

#---------------------------------     DEFINE PATHS AND FILE NAMES:  ----------------------------------

#--- we need to define some directory variables to use rmarkdown from snakemake
#---                !!!! IMPORTANT !!!!
#--- ! paths inside input and output should be relative
#--- ! if params are passed to Rmarkdown document, paths to files or folders should be absolute   


BASEDIR = workflow.basedir + "/"                    #--- the base dir (where Snakefile is) 
WORKDIR = os.getcwd() + "/"                         #--- current work dir (important for rmarkdown)

SCRIPTDIR   = BASEDIR + "scripts/"                  #--- scripts dir (relative to snakefile)
TEMPLATEDIR = BASEDIR + "report_templates/"         #--- template dir (relative to snakefile)

#--- pathout is ignored, since all paths are relative from either base or workdir
## notes:
## maybe set pathout by setting working dir in snakemake call
# PATHOUT      = config["PATHOUT"]
# LOGS        = config["LOG"] 



#--- the in/output directories can be renamed easily
#--- path to bam files is absolute, such that I can run workflow in any other directory
BAMDIR      = '06_sorted/'  # BASEDIR + "test_dataset/sample_bam_files/" 


#--- output paths have to be relative to working directory, 
#--- such that I can run workflow in any other directory and
#--- it always sets up the same structure
METHCALLDIR = "methylation_calls/"
SEGDIR      = "segmentation/"
ANNODIR     = "annotation/"
FINALDIR    = "final_Report/"




## Bam processing
rule bam_methCall:
    input:  
        template    = TEMPLATEDIR + "methCall.report.Rmd",
        bamfile     = BAMDIR  + "{prefix}.sorted.bam"
    output: 
        report      = METHCALLDIR + "{prefix}.sorted_meth_calls.nb.html",
        rdatafile   = METHCALLDIR + "{prefix}.sorted_methylRaw.RData",
        callFile    = METHCALLDIR + "{prefix}.sorted_CpG.txt",
        knitr_meta  = touch(FINALDIR + "{prefix}/knitr_meta_notuse.rds")
    params:
        ## absolute path to bamfiles
        inBam       = WORKDIR + BAMDIR + "{prefix}.sorted.bam",
        assembly    = config["GENOME_VERSION"],
        mincov      = int(config["bam_methCall_args_mincov"]), ##TODO
        minqual     = int(config["bam_methCall_args_minqual"]), ##TODO
        ## absolute path to output folder in working dir
        rdata       = WORKDIR + METHCALLDIR + "{prefix}.sorted_methylRaw.RData"
    log:
        METHCALLDIR + "{prefix}.sorted_meth_calls.log"
    message:
        "Processing of bam file:\n" 
        "   input     : {input.bamfile}" + "\n"
        "Generating:"+ "\n" 
        "   report    : {output.report}" + "\n" 
        "   rdata     : {output.rdatafile}" + "\n" 
        "   methCalls : {output.callFile}"
    script:
        SCRIPTDIR + "report_functions.R"  




## Segmentation
rule methseg:
    ## paths inside input and output should be relative
    input:  
        template    = TEMPLATEDIR + "methseg.report.Rmd",
        rdatafile   = METHCALLDIR + "{prefix}.sorted_methylRaw.RData"
    output: 
        report      = SEGDIR + "{prefix}.sorted_meth_segments.nb.html",
        grfile      = SEGDIR + "{prefix}.sorted_meth_segments_gr.RData",
        bedfile     = SEGDIR + "{prefix}.sorted_meth_segments.bed",
        knitr_meta  = touch(FINALDIR + "{prefix}/knitr_meta_notuse.rds")    
    params:
        rdata       = WORKDIR + METHCALLDIR + "{prefix}.sorted_methylRaw.RData",
        grdata      = WORKDIR + SEGDIR + "{prefix}.sorted_meth_segments_gr.RData",
        outBed      = WORKDIR + SEGDIR + "{prefix}.sorted_meth_segments.bed"
    log:
        SEGDIR + "{prefix}.sorted_meth_segments.log"
    message:
        "Segmentation of sample file:\n"
        "   input     : {input.rdatafile}" + "\n" 
        "Generating:"+ "\n"
        "   report    : {output.report}" + "\n"  
        "   grfile    : {output.grfile} " +"\n" 
        "   bedfile   : {output.bedfile}" +"\n"

    script:
        SCRIPTDIR + "report_functions.R" 


## Aquisition of gene features
rule fetch_refGene:
    output: refgenes = ANNODIR + "refseq.genes.{assembly}.bed"
    params: 
        assembly = "{assembly}"
    log:
        ANNODIR + "fetch_refseq.genes.{assembly}.log"
    message:
        "Fetching RefSeq genes for Genome assembly: {wildcards.assembly}"
    script:  
        SCRIPTDIR + "fetch_refGene.R" 


## Annotation with gene features
rule methseg_annotation:
    input:  
        template    = TEMPLATEDIR + "annotation.report.Rmd",
        bedfile     = SEGDIR + "{prefix}.sorted_meth_segments.bed",
        refgenes    = ANNODIR + "refseq.genes.{assembly}.bed"
    output: 
        ## touch this imaginary file, to get the directory
        knitr_meta  = touch(FINALDIR + "{prefix}/knitr_meta_{assembly}_notuse.rds"),
        report      = ANNODIR + "{prefix}.sorted_{assembly}_annotation.nb.html",
        # knitr_meta  = FINALDIR + "{prefix}/knitr_meta_{assembly}.rds"
        #knitr_meta  = lambda wildcards: pass_knitr_meta(FINALDIR,wildcards.prefix)
    params:
        inBed       = WORKDIR + SEGDIR + "{prefix}.sorted_meth_segments.bed",
        assembly    = "{assembly}",# expand(config["reference"]),
        refseqfile  = WORKDIR + ANNODIR + "refseq.genes.{assembly}.bed"
    log:
        ANNODIR + "{prefix}.sorted_{assembly}_annotation.log"
    message:
        "Annotation of Segments:\n"
        "   input     : {input.bedfile}" + "\n"
        "Generating:" + "\n"
        "   report    : {output.report}"
    script:
        SCRIPTDIR + "report_functions.R"   



### note that Final report can only be generated 
### if one of the intermediate has been genereted,
### so make sure that at least one has been run already
### right now ensured with 'rules.methseg_annotation.output' as input
### 

# rule multiqc:
#     input:
#         workdir = WORKDIR + "."


# def get_fastq_name(full_name):
#     print(full_name)
#     if "se_bt2" in full_name:
#         return full_name[0:full_name.find("se_bt2")-1] #"Hello "+ full_name# #
#     elif "val_1_bt2" in full_name:
#         return [full_name[0:full_name.find("val_1_bt2")-2]+i for i in i=[1 2]] # "Hi "+ full_name#[]
#     else:
#         raise Exception("=== ERROR: Cannot identify fastq_name. STOP! ===")

## Final Report
rule final_report:
    input:  
        rules.methseg_annotation.output,
        index       = TEMPLATEDIR + "index.Rmd",   
        references  = TEMPLATEDIR + "references.Rmd",
        sessioninfo = TEMPLATEDIR + "sessioninfo.Rmd"
        #fastq_name  = lambda wildcards: touch(get_fastq_name(wildcards.prefix))#,
        ## add rawqc at some point
    output: 
        finalreport = FINALDIR + "{prefix}.sorted_{assembly}_final.nb.html",
    params:
        finalreportdir = WORKDIR + FINALDIR + "{prefix}/",
        workdir = WORKDIR#,
        # fastq_name = lambda wildcards: get_fastq_name(wildcards.prefix)#,
        # rules.fastqc_raw.output
    log:
        FINALDIR + "{prefix}.sorted_{assembly}_final.log"
    message:
        "Compiling Final Report:\n"
        "   report    : {output.finalreport}."#\n{params.fastq_name}"
        
    script:
        SCRIPTDIR + "multireport_functions.R"  




